### PR TITLE
Add useListenableMap

### PR DESCRIPTION
**Pull request by ronnieeeeee**
Hi @rrousselGit 😄
I developed useListenableMap() [suggested](https://github.com/rrousselGit/flutter_hooks/issues/287#issuecomment-1061818631) by Remi.

```dart
class Example extends HookWidget {
  @override
  Widget build(BuildContext context) {
   final listenable = useTextEditingController();
   final bool textIsEmpty =
        useListenableMap(listenable, () => listenable.text.isEmpty);
   return Column(
      children: [
        TextField(controller: listenable),
        ElevatedButton(
            // if textIsEmpty is false,　the button is disabled.
            onPressed: textIsEmpty ? null : () => print("Pressed!"),
            child: Text("Button")),
      ],
    );
  }
}
```

### About useListenableMap()
Subscribe to a Listenable, when [Listenable] update execute callback(). the result is the return value.
It doesn't rebuild the widget unless that return of callback() result does not change. 
An unnecessary rebuild does not occur so do not unnecessary performance.

There was a[ problem](https://github.com/rrousselGit/flutter_hooks/issues/287) that writing the same process without useListenableMap would be verbose, so I created it.

### CODE REVIEW

**Comment by rrousselGit** [comment]
Thanks, awesome work 😄 

I don't like the name though. It's not really a "map".
Maybe useListenableSelector?

File: packages/flutter_hooks/lib/src/listenable_map.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+///         ElevatedButton(
+///             // if textIsEmpty is false,　the button is disabled.
+///             onPressed: textIsEmpty ? null : () => print("Pressed!"),
+///             child: Text("Button")),
+///       ],
+///     );
+///   }
+/// }
+/// ```
+R useListenableMap<T extends Listenable, U extends R Function(), R>(
```
Comment (rrousselGit) [inline packages/flutter_hooks/lib/src/listenable_map.dart:28]: That's too many generics.

Only R should be necessary, the others shouldn't be here.

File: packages/flutter_hooks/lib/src/listenable_map.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  final U callback;
+
+  @override
+  _ListenableMapHookState<T, U, R> createState() =>
+      _ListenableMapHookState<T, U, R>();
+}
+
+class _ListenableMapHookState<T extends Listenable, U extends R Function(), R>
+    extends HookState<ValueNotifier<R>, _ListenableMapHook<T, U, R>> {
+  late final ValueNotifier<R> _state = ValueNotifier<R>(hook.callback())
```
Comment (rrousselGit) [inline packages/flutter_hooks/lib/src/listenable_map.dart:47]: Don't use a ValueNotifier. Use a property and call setState on change

File: packages/flutter_hooks/test/use_listenable_map_test.dart [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    listenable.value++;
+    await tester.pump();
+    expect(listenable.value, 46);
+    expect(isOdd, false);
+    expect(element.dirty, false);
+
+    await tester.pumpWidget(const SizedBox());
+    // ignore: invalid_use_of_protected_member
+    expect(() => listenable.hasListeners, throwsFlutterError);
```
Comment (rrousselGit) [inline packages/flutter_hooks/test/use_listenable_map_test.dart:79]: Why should this line be throwing?

The hook listenable isn't disposed, so calling hasListeners shouldn't throw

File: packages/flutter_hooks/test/use_listenable_map_test.dart [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+void main() {
+  testWidgets('debugFillProperties', (tester) async {
+    await tester.pumpWidget(HookBuilder(
+      builder: (context) {
+        final listenable = ValueNotifier<int>(42);
+        useListenableMap<ValueNotifier<int>, bool Function(), bool>(
+            listenable, () => listenable.value.isOdd);
+        return const SizedBox();
+      },
+    ));
```
Comment (rrousselGit) [inline packages/flutter_hooks/test/use_listenable_map_test.dart:17]: Nit: add trailing comma and format

File: packages/flutter_hooks/test/use_listenable_map_test.dart [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import 'package:flutter_hooks/src/hooks.dart';
+import 'mock.dart';
+
+void main() {
+  testWidgets('debugFillProperties', (tester) async {
+    await tester.pumpWidget(HookBuilder(
+      builder: (context) {
+        final listenable = ValueNotifier<int>(42);
+        useListenableMap<ValueNotifier<int>, bool Function(), bool>(
+            listenable, () => listenable.value.isOdd);
```
Comment (rrousselGit) [inline packages/flutter_hooks/test/use_listenable_map_test.dart:14]: Add trailing comma

File: packages/flutter_hooks/test/use_listenable_map_test.dart [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      equalsIgnoringHashCodes(
+        'HookBuilder\n'
+        ' │ useListenableMap<ValueNotifier<int>>:\n'
+        ' │   ValueNotifier<bool>#00000(false)\n'
+        ' └SizedBox(renderObject: RenderConstrainedBox#00000)\n'
+        '',
+      ),
+    );
+  });
+  testWidgets('useListenableMap', (tester) async {
```
Comment (rrousselGit) [inline packages/flutter_hooks/test/use_listenable_map_test.dart:34]: Add empty line before test

File: packages/flutter_hooks/lib/src/listenable_map.dart (hunk 1 above)
Comment (ronnieeeeee) [reply packages/flutter_hooks/lib/src/listenable_map.dart:28]: > Thanks, awesome work 😄
> 
> I don't like the name though. It's not really a "map". Maybe useListenableSelector?

Thank you, Remi 😄
I'm so glad to hear you say that!
I prefer useListenableSelector as naming. I am rethinking the naming.

Thank you for your many suggestions for my code, even though you seem to be busy. I will fix it.

File: packages/flutter_hooks/test/use_listenable_map_test.dart (hunk 4 above)
Comment (ronnieeeeee) [reply packages/flutter_hooks/test/use_listenable_map_test.dart]: I have too much knowledge to learn, so I will study. Thank you, Remi 👍

**Comment by ronnieeeeee** [comment]
> 

I couldn't think of good naming, so to name useListenableSelector.

File: packages/flutter_hooks/lib/src/listenable_selector.dart [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    hook.listenable.addListener(_listener);
+  }
+
+  @override
+  R build(BuildContext context) => _state;
+
+  void _listener() {
+    final result = hook.callback();
+    if (_state != result) {
+      _state = result;
```
[the remark this answers is not in the document]
Comment (ronnieeeeee) [reply packages/flutter_hooks/lib/src/listenable_selector.dart:59]: How embarrassing!
I was referring only to primitive.dart and listenable.dart.
So without thinking, I assumed setState would be used like this `setState(() {});`.

**Comment by rrousselGit** [comment]
Oh, I spotted a few problems. 🙈 
I'll revert the PR. It'd be great if you can have a look @ronnieeeeee 

The problems being:
- we need to handle `listenable` changing
- similarly, when the widget rebuilds, we need to re-invoke the callback to update the `_state`
  (because users could write `useListenableSelector(listenable, () => listenable.value + widget.param)`

**Comment by ronnieeeeee** [comment]
Hi Remi(@rrousselGit), I am sorry that the implementation of the useListeanbleSelector did not work,
even though you took time out of your busy responding to other issues and PRs.
And thank you for all the fixes and suggestions.
However, I am very interested in flutter_hooks (and riverpod) and enjoy being involved in its development, 
so I will continue to implement useListeanbleSelector. 👍

**Comment by ronnieeeeee** [comment]
I am thinking of adding to useListenableSelector the functionality that useEffect and others implement by taking keys as the second argument.

The plan looks like this.
```dart
useListenableSelector(listenable, () => listenable.value + widget.param, [widget.param])
useListenableSelector(listenable, () => listenable.value + widget.param, [listenable,widget.param])
```

**Comment by ronnieeeeee** [comment]
Or consider also useListenableReducer.

**Comment by rrousselGit** [comment]
> I am thinking of adding to useListenableSelector the functionality that useEffect and others implement by taking keys as the second argument.
> 
> The plan looks like this.
> 
> ```dart
> useListenableSelector(listenable, () => listenable.value + widget.param, [widget.param])
> useListenableSelector(listenable, () => listenable.value + widget.param, [listenable,widget.param])
> ```

I don't think this make sense for this hook.

I'd expect to simply override `didUpdateHook` and handle changes

**Comment by rrousselGit** [comment]
Selectors typically don't act as a cache. It's fine to call the callback again when the hook rebuilds

If folks want a cache, they should use useMemoized

**Comment by ronnieeeeee** [comment]
> 

That makes sense、I'll try it.

**Comment by ronnieeeeee** [comment]
> Selectors typically don't act as a cache. It's fine to call the callback again when the hook rebuilds

I surely agree.
Thank you for the advice.

**Pull request merged**